ops/model.py

```diff
@@ -797,6 +927,267 @@ def __init__(self, name: str, address_info: '_AddressDict'):
+            raise RuntimeError(
+                'cannot revoke secret access from application that is not remote to the specified relation')
+
+        relation = relation or target
+        assert isinstance(relation, Relation)  # type guard
+
+        unit = target.name if isinstance(target, Unit) else None
+        self._backend.secret_revoke(self.id, relation.id, unit_id=unit)
+
+    def prune(self):
+        """Remove this secret revision from juju.
+
+        This should be called while handling `secret-remove` events, otherwise
+        you cannot be ensured that this revision is not still being consumed by
+        some remote units.
+        """
+        if not self._am_owner:
+            raise RuntimeError(
+                'cannot prune secret {} which is owned by a different unit'.format(self))
+        if self._revision is None:
+            raise RuntimeError(
+                'secret {} has no revision to prune - are you outside a secret-remove hook call?')
+        self._backend.secret_remove(self.id, revision=self._revision)
+
+    def remove(self):
+        """Remove all revisions of this secret from juju, effectively deleting it.
+
+        This should typically happen when the relation the secret is bound to is broken.
+        Removing the secret prematurely means that all units/applications it has
+        been granted to might attempt to get the secret
+        """
+
+        if not self._am_owner:
+            raise RuntimeError(  # todo unify with SecretOwnershipError
+                'cannot remove secret {} which is owned by a different unit'.format(self))
+        self._backend.secret_remove(self.id)
+
+    def update(self) -> '_Secret':
+        """Fetch the latest available revision of this secret.
+
+        Returns a Secret object pointing to the newest available revision.
+        The old Secret revision is not updated in-place, so it will keep pointing
+        to the old revision.
+
+        Examples:
+            >>> def _on_update_status(self: ops.charm.CharmBase, _):
+            >>>     old_revision = self.model.get_secret('my_label')
+            >>>     new_revision = old_revision.update()
+            >>>     if new_revision != old_revision:
+            >>>         logger.info('updated to new revision!')
+        """
+        if self._am_owner:
+            raise RuntimeError(
+                'cannot update secret {} which is owned by this unit - did you mean to call Secret.set?'.format(
+                    self))
+        self._backend.secret_get(self.id, update=True)
+
+        return _Secret(backend=self._backend,
+                       secret_id=self.id,
+                       label=self.label,
+                       revision=None,  # cannot know the revision as a non-owner
+                       am_owner=self._am_owner)
+
+    def get(self, key: str,
+            update: bool = False,
+            peek: bool = False) -> str:
```

Comment:
We tried to keep the signatures close to the hook tools' ones, but I see how it could be confusing here.

Since we have a separate `update` method already, we could say:
each `_Secret` object can be used to interact with a specific revision only.
if you want to update to a new revision, the only way is by calling `_Secret.update()` which will give you back a pointer to the new revision.

We add a new `.peek(key:str)` method which allows you to preview a new secret's contents without updating just yet.

We remove both `peek` and `update` from `get`'s signature: `get` can only be used to get a value for the currently tracked revision.

Reply:
closing this as the same conversation is going on above.
@rbarry82 can you read my and @wallyworld 's comments and update with your feedback?